Add a PUBLISH message for publisher initiated subscriptions (#995)
moq-wg/moq-transport: draft-ietf-moq-transport.md

window upon detecting loss.  In both cases, the large reduction in sending rate might
cause issues with latency sensitive applications.

# Publishing and Retrieving Tracks

## Subscriptions

A subscription can be initiated by either a publisher or a subscriber.  A
publisher initiates a subscription to a track by sending the PUBLISH message.
The subscriber either accepts or rejects the subscription using PUBLISH_OK or
PUBLISH_ERROR.  A subscriber initiates a subscription to a track by sending the
SUBSCRIBE message.  The publisher either accepts or rejects the subscription
using SUBSCRIBE_OK or SUBSCRIBE_ERROR.  Once either of these sequences is
successful, the subscription can be updated by the subscriber using
SUBSCRIBE_UPDATE, terminated by the subscriber using UNSUBSCRIBE, or terminated
by the publisher using SUBSCRIBE_DONE.

All subscriptions have a Forward State which is either 0 or 1.  If the Forward
State is 0, the publisher does not send objects for the subscription.  If the
Forward State is 1, the publisher sends objects.  The initiator of the
subscription sets the initial Forward State in either PUBLISH or SUBSCRIBE.  The

> sender of PUBLISH_OK can update the Forward State based on its preference.  Once
> established, the subscriber can change the Forward State by sending
> SUBSCRIBE_UPDATE.

Reviewer: ```suggestion
sender of PUBLISH_OK can update the Forward State based on its preference.  Once
the subscription is established, the subscriber can update the Forward State by sending
SUBSCRIBE_UPDATE.
```

Either endpoint can initiate a subscription to a track without exchanging any
prior messages other than SETUP.  Relays MUST NOT send any PUBLISH messages
without knowing the client is interested in and authorized to receive the

> content. The communication of intent and authorization can be accomplished by

Reviewer 2: ```
Relays MUST NOT send any PUBLISH messages
without knowing the client is interested in and authorized to receive the
content. 
```
This is unenforceable and should not be normative. Bad relays will do this and we should provide mechanisms for the receivers to protect themselves. Additionally, Publish should carry a token to enforce authorization.

Reviewer 3: Agreed, given this is unenforceable, I don't think we can go stronger than SHOULD

Reviewer 4: Not so much about the issue here but I strongly object to this idea that things that are not enforceable can not be normative. We need to tell relays what they can and can not do. This is also very testable and observable by a client.

> the client sending SUBSCRIBE_ANNOUNCES, or conveyed in other mechanisms out of

Reviewer: ```suggestion
the client sending SUBSCRIBE_ANNOUNCES/SUBSCRIBE message, or conveyed in other mechanisms out of
```

Author: I don't think we want to send PUBLISH in response to SUBSCRIBE.  We would send SUBSCRIBE_OK plus track data.

forward the result to the application, so the application can decide which other
publishers to contact, if any.

An UNSUBSCRIBE_ANNOUNCES withdraws a previous SUBSCRIBE_ANNOUNCES. It does not
prohibit original publishers from sending further ANNOUNCE or PUBLISH messages,

> but relays MUST not send a PUBLISH message to a client without knowing the

Reviewer: ```suggestion
but relays MUST not send any subsequent PUBLISH message to a client without knowing the
```

> but relays MUST not send a PUBLISH message to a client without knowing the
> client is interested in and authorized to receive the content.

Author: ```suggestion
but relays MUST not send any further PUBLISH messages to a client without
knowing the client is interested in and authorized to receive the content.
```

## Publisher Interactions

There are two ways to publish through a relay - sending a PUBLISH messages for a
specific Track or sending an ANNOUNCE message for a Track Namespace and waiting
for SUBSCRIBE.  Relays MAY respond to PUBLISH with PUBLISH_OK in Forward State=0
until there are known subscribers for new tracks. ANNOUNCE enables the relay to

> know which publisher to forward a SUBSCRIBE to.

Reviewer: ```suggestion
There are two ways to publish through a relay - sending a PUBLISH messages for a
specific Track or sending an ANNOUNCE message for a Track Namespace and waiting
for SUBSCRIBE.  Relays MAY respond to PUBLISH with PUBLISH_OK in Forward State=0
until there are known subscribers for new tracks. ANNOUNCE enables the relay to
know which publisher to forward a SUBSCRIBE to.```

There are two methods to publish through a relay:
1. Sending a PUBLISH message for a specific track. Relays MAY respond to PUBLISH with PUBLISH_OK with Forward State=0 until subscribers for new tracks are identified.
2. Sending an ANNOUNCE message for a track namespace and waiting for SUBSCRIBE(s) to the tracks in that namespace. This allows the relay to determine which publisher(s) to forward the SUBSCRIBE(s) to.

> There are two ways to publish through a relay - sending a PUBLISH messages for a
> specific Track or sending an ANNOUNCE message for a Track Namespace and waiting
> for SUBSCRIBE.  Relays MAY respond to PUBLISH with PUBLISH_OK in Forward State=0
> until there are known subscribers for new tracks. ANNOUNCE enables the relay to
> know which publisher to forward a SUBSCRIBE to.

Reviewer: ```suggestion
There are two methods to publish through a relay:

1. Sending a PUBLISH message for a specific track. Relays MAY respond to PUBLISH with
PUBLISH_OK and Forward State=0, until subscribers for new tracks are identified.

2.Sending an ANNOUNCE message for a track namespace and waiting for SUBSCRIBE(s) to the track within the announced track namespace. This allows the relay to determine which publisher(s) to forward the SUBSCRIBE(s) to.```

> Relays MUST verify that publishers are authorized to publish the content
> associated with the set of tracks whose Track Namespace matches the namespace in
> an ANNOUNCE, or the Full Track Name in PUBLISH. Where the authorization and
> identification of the publisher occurs depends on the way the relay is managed
> and is application specific.

Reviewer: ```suggestion
Relays MUST verify that publishers are authorized to publish the content
associated with the set of tracks whose Track Namespace matches the namespace in
an ANNOUNCE, or the Full Track Name in PUBLISH. Where the authorization and
identification of the publisher occurs depends on the way the relay is managed
and is application specific.
```
There are two methods to publish through a relay:

1. Sending a PUBLISH message for a specific track. Relays MAY respond to PUBLISH with 
PUBLISH_OK and Forward State=0, until subscribers for new tracks are identified.

2. Sending an ANNOUNCE message for a track namespace and waiting for SUBSCRIBE(s) to the track within the announced track namespace. This allows the relay to determine which publisher(s) to forward the SUBSCRIBE(s) to.

## Publisher Interactions

There are two ways to publish through a relay:

1. Send a PUBLISH message for a specific Track to the relay. The relay MAY
respond with PUBLISH_OK in Forward State=0 until there are known subscribers for
new tracks.

2. Send an ANNOUNCE message for a Track Namespace to the relay. This enables the
relay to send SUBSCRIBE messages for Tracks in this Namespace in response to

> received SUBSCRIBE messages.

Reviewer: Send an ANNOUNCE message for a Track Namespace to the relay. This enables the
relay to send SUBSCRIBE messages **to the publishers** for Tracks in this Namespace in response to the received SUBSCRIBE messages.

> relay to send SUBSCRIBE messages for Tracks in this Namespace in response to
> received SUBSCRIBE messages.

Author: ```suggestion
relay to send SUBSCRIBE messages to publishers for Tracks in this Namespace in
response to received SUBSCRIBE messages.
```

and is application specific.

A Relay can receive announcements for the same Track Namespace or PUBLISH
messages for the same Track from multiple publishers and is expected to treat
them uniformly.

When a publisher wants to stop new subscriptions for an announced namespace it
sends an UNANNOUNCE. A subscriber indicates it will no longer subcribe to tracks
in a namespace it previously responded ANNOUNCE_OK to by sending an
ANNOUNCE_CANCEL.

A relay manages sessions from multiple publishers and subscribers, connecting
them based on the Track Namespace or Full Track Name.  Prefix matching is used
to determine which publishers receive a SUBSCRIBE or which subscribers receive a
PUBLISH. For example, a SUBSCRIBE namespace=(foo,bar), track=x message will be
forwarded to the sessions that sent ANNOUNCE namespace=(foo) and ANNOUNCE
namespace=(foo, bar) respectively, but not one that sent ANNOUNCE
namespace=(foobar).  Relays MUST forward SUBSCRIBE messages to all publishers
and ANNOUNCE and PUBLISH messages to all subscribers that have a namespace
prefix match.

When a relay receives an incoming SUBSCRIBE that triggers an upstream

> subscription, it SHOULD send a SUBSCRIBE request to each publisher that has

Reviewer: ```suggestion
subscription, it MUST send a SUBSCRIBE request to each publisher that has
```

Author: This SHOULD was there before.

Reviewer: I understand that and it will be nice if we can update it

Author: I filed #1037 so we can track that separately from this PR, which is introducing PUBLISH mechanics.

announced the subscription's namespace or prefix thereof, unless it already has
an active subscription for the Objects requested by the incoming SUBSCRIBE
request from all available publishers.  If it already has a matching upstream

> subscription in Forward State=0, it SHOULD send a SUBSCRIBE_UDPATE with

Reviewer: ```suggestion
subscription in Forward State=0, it MUST send a SUBSCRIBE_UDPATE with
```

Author: This is for consistency with how we define Forward for SUBSCRIBE - I think it says the relay SHOULD send Forward=1 upstream.

Forward=1 to all publishers.

> When a relay receives an incoming PUBLISH message, it SHOULD send a PUBLISH

Reviewer: ```suggestion
When a relay receives an incoming PUBLISH message, it MUST send a PUBLISH
```

request to each subscriber that has subscribed (via SUBSCRIBE_ANNOUNCES)
to the track's namespace or prefix thereof.

When a relay receives an incoming ANNOUNCE for a given namespace, for

> each active upstream subscription that matches that namespace, it SHOULD send a

Reviewer: ```suggestion
each active upstream subscription that matches that namespace, it MUST send a
```

SUBSCRIBE to the publisher that sent the ANNOUNCE.  When it receives an incoming

> PUBLISH message for a track that has active subscribers, it SHOULD respond

Reviewer: ```suggestion
PUBLISH message for a track that has active subscribers, it MUST respond
```

Author: Keeping forward state normative language matching SUBSCRIBE here

> PUBLISH message for a track that has active subscribers, it SHOULD respond

Reviewer 3: By active, we mean that at least one subscription has a Forward State=1?

~~~
PUBLISH Message {
  Type (i) = 0x6,
  Length (i),
  Request ID (i),
  Track Namespace (tuple),
  Track Name Length (i),
  Track Name (..),
  Track Alias (i),
  Group Order (8),
  ContentExists (8),
  [Largest (Location),]
  Forward (8),
  Number of Parameters (i),
  Parameters (..) ...,
}
~~~
{: #moq-transport-publish-format title="MOQT PUBLISH Message"}

* Request ID: See {{request-id}}.

* Track Namespace: Identifies a track's namespace as defined in ({{track-name}})

* Track Name: Identifies the track name as defined in ({{track-name}}).

* Track Alias: The identifer used for this track in Subgroups or Datagrams (see
  {{track-alias}}). The same Track Alias MUST NOT be used to refer to two
  different Tracks simultaneously. If a subscriber receives a PUBLISH that
  uses the same Track Alias as a different track with an active subscription, it
  MUST close the session with error 'Duplicate Track Alias'.

* ContentExists: 1 if an object has been published on this track, 0 if not.
If 0, then the Largest Group ID and Largest Object ID fields will not be
present. Any other value is a protocol error and MUST terminate the

> session with a Protocol Violation ({{session-termination}}).

Reviewer: I am not sure if we need ContentExists for publish message, since the intent of publish message is to publish content.

Author: I think we do.  I imagine in most cases, Content Exists = 0 (hence no Largest Location yet).  However, if a publisher is starting in the middle of a track (reconnect?) then Content Exists = 1 and Largest Location will be present.

  is a Protocol Violation.  0 indicates the publisher will not transmit any
  objects until the subscriber sets the Forward State to 1. 1 indicates the
  publisher will start transmitting objects immediately, even before PUBLISH_OK.

* Parameters: The parameters are defined in {{version-specific-params}}.

## PUBLISH_OK {#message-publish-ok}

The subscriber sends an PUBLISH_OK control message to acknowledge the successful
authorization and acceptance of a PUBLISH message, and establish a subscription.

~~~
PUBLISH_OK Message
{
  Type (i) = 0x7,
  Length (i),
  Request ID (i),
  Forward (8),
  Subscriber Priority (8),
  Group Order (8),
  Filter Type (i),
  [Start (Location)],
  [EndGroup (i)],
  Number of Parameters (i),
  Parameters (..) ...,
}
~~~
{: #moq-transport-publish-ok format title="MOQT PUBLISH_OK Message"}

* Request ID: The Request ID of the PUBLISH this message is replying to
  {{message-publish}}.

* Forward: The Forward State for this subscription, either 0 (don't
  forward) or 1 (forward).

* Subscriber Priority: The Subscriber Priority for this subscription.

* Group Order: The Group Order preference for this subscription.  This

> overwrites the GroupOrder specified PUBLISH.

Reviewer 5: If a subscriber receives a PUBLISH with GroupOrder == Ascending, but overrides it to Descending, what does the peer do it it is unwilling or unable to serve in that order? Does it just immediately send a SUBSCRIBE_DONE with an error code?

Author: Group Order in a subscription is only input to prioritization, not object production, so I don't think there's a case where the publisher is unwilling/unable.

* Filter Type, Start, End Group: See {{message-subscribe-req}}.

* Parameters: Parameters associated with this message.

## PUBLISH_ERROR {#message-publish-error}

The subscriber sends an PUBLISH_ERROR control message for to reject
a subscription initiated by PUBLISH.

~~~
PUBLISH_ERROR Message
{
  Type (i) = 0x8,
  Length (i),
  Request ID (i),
  Error Code (i),

>   Reason Phrase Length (i),
>   Reason Phrase (..),

Author: ```suggestion
  Error Reason (Reason Phrase),
```

}
~~~
{: #moq-transport-publish-error format title="MOQT PUBLISH_ERROR Message"}

* Request ID: The Request ID of the PUBLISH this message is replying to
  {{message-publish}}.

* Error Code: Identifies an integer error code for failure.

> * Reason Phrase: Provides the reason for error.

Author: ```suggestion
* Error Reason: Provides the reason for subscription error. See {{reason-phrase}}.
```

  subscriber exceeds its implementation defined limit.

## PUBLISH {#message-publish}

The publisher sends the PUBLISH control message to initiate a subscription to a
track. The receiver verifies the publisher is authorized to publish this track.

~~~
PUBLISH Message {
  Type (i) = 0x6,
  Length (i),
  Request ID (i),
  Track Namespace (tuple),
  Track Name Length (i),
  Track Name (..),
  Track Alias (i),
  Group Order (8),
  ContentExists (8),
  [Largest (Location),]
  Forward (8),
  Number of Parameters (i),
  Parameters (..) ...,
}
~~~
{: #moq-transport-publish-format title="MOQT PUBLISH Message"}

* Request ID: See {{request-id}}.

* Track Namespace: Identifies a track's namespace as defined in ({{track-name}})

* Track Name: Identifies the track name as defined in ({{track-name}}).

* Track Alias: The identifer used for this track in Subgroups or Datagrams (see
  {{track-alias}}). The same Track Alias MUST NOT be used to refer to two
  different Tracks simultaneously. If a subscriber receives a PUBLISH that
  uses the same Track Alias as a different track with an active subscription, it
  MUST close the session with error 'Duplicate Track Alias'.

> 

Reviewer 6: Group Order is missing from this text.

* Forward: The forward mode for this subscription.  Any value other than 0 or 1
  is a Protocol Violation.  0 indicates the publisher will not transmit any
  objects until the subscriber sets the Forward State to 1. 1 indicates the
  publisher will start transmitting objects immediately, even before PUBLISH_OK.

* Parameters: The parameters are defined in {{version-specific-params}}.

## PUBLISH_OK {#message-publish-ok}

The subscriber sends an PUBLISH_OK control message to acknowledge the successful
authorization and acceptance of a PUBLISH message, and establish a subscription.

~~~
PUBLISH_OK Message
{
  Type (i) = 0x7,
  Length (i),
  Request ID (i),
  Forward (8),
  Subscriber Priority (8),
  Group Order (8),
  Filter Type (i),
  [Start (Location)],
  [EndGroup (i)],
  Number of Parameters (i),
  Parameters (..) ...,
}
~~~
{: #moq-transport-publish-ok format title="MOQT PUBLISH_OK Message"}

* Request ID: The Request ID of the PUBLISH this message is replying to
  {{message-publish}}.

* Forward: The Forward State for this subscription, either 0 (don't
  forward) or 1 (forward).

* Subscriber Priority: The Subscriber Priority for this subscription.

> * Group Order: The Group Order preference for this subscription.  This

Reviewer 3: Future-looking comment: This seems like it'd be better as a param, because I would think that typically you wouldn't want to change GroupOrder?

> * Group Order: The Group Order preference for this subscription.  This

Reviewer 6: This whole time I assumed (based on the word "request") that the Group Order in SUBSCRIBE was advisory, but I see now how one could read it as mandatory.

I agree that the current PR is consistent with the ID that the subscriber always chooses. I can't say if that's wise or not.

overwrites the GroupOrder specified PUBLISH.

* Filter Type, Start, End Group: See {{message-subscribe-req}}.

* Parameters: Parameters associated with this message.

## PUBLISH_ERROR {#message-publish-error}

> The subscriber sends an PUBLISH_ERROR control message for to reject

Reviewer 3: ```suggestion
The subscriber sends an PUBLISH_ERROR control message to reject
```

* Parameters: The parameters are defined in {{version-specific-params}}.

The publisher will respond with SUBSCRIBE_ANNOUNCES_OK or
SUBSCRIBE_ANNOUNCES_ERROR.  If the SUBSCRIBE_ANNOUNCES is successful, the
publisher will immediately forward existing ANNOUNCE and PUBLISH messages that
have the Track Namespace Prefix.  If the set of matching ANNOUNCE messages
changes, the publisher sends the corresponding ANNOUNCE or UNANNOUNCE
message.

A subscriber cannot make overlapping namespace subscriptions on a single
session.  Within a session, if a publisher receives a SUBSCRIBE_ANNOUNCES with a
Track Namespace Prefix that is a prefix of an earlier SUBSCRIBE_ANNOUNCES or
vice versa, it MUST respond with SUBSCRIBE_ANNOUNCES_ERROR, with error code

> Namespace Prefix Overlap.

Reviewer 6: I see that now. Meh. I don't have a suggested rewording except to just do another sentence that just states the reverse.

Maybe it's just a common sense reading, but I'd prefer it if it was an earlier, **active** SUBSCRIBE_ANNOUNCES rather than just an earlier one.

0x03 and Extension Headers Length is 0, it MUST close the session with Protocol
Violation.

> * Track Alias: The Track Alias ({{track-alias}} indicating the track this
>   Datagram belongs to.  If an endpoint receives a datagram with an unknown Track
>   Alias, it MAY drop the datagram or choose to buffer it for a brief period to
>   handle reordering with the control message that establishes the Track Alias.
> 

Author: ```suggestion
```

If an endpoint receives a subgroup with an unknown Track Alias, it MAY abandon
the stream, or choose to buffer it for a brief period to handle reordering with
the control message that establishes the Track Alias.  The endpoint MAY withhold
stream flow control beyond the SUBGROUP_HEADER until the Track Alias has been

> established.  To prevent deadlocks, the publisher MUST allocate connection flow

Reviewer 4: not a request for any changes but just a question .... in other places we said 

```
Given the critical nature of control messages and their relatively
small size, the control stream SHOULD be prioritized higher than all
subscribed Objects.
```

Does that prioritize higher imply this flow control as well ? I sort of feel like all of this should be covered in the {#session-init} or some other place.

Author: This text was part of #977, not this PR.

> Does that prioritize higher imply this flow control as well ?

I don't think it's guaranteed, though it's quite sensible.  SHOULD vs MUST also needs to be resolved.

>  sort of feel like all of this should be covered in the {#session-init} or some other place.

I added it with Track Alias since that introduced the deadlock condition, but we can move it around.
